### PR TITLE
mvcc: avoid extra allocations in TxnEnd

### PR DESCRIPTION
**Pull request by nekto0n**
Investigating #4816 found that extra allocations are happening while looking for watchers (by key or by range). Tried to avoid by explicitly checking if there are any.

Running new benchmark (not quite sure if it is correct).
Before:
```
BenchmarkWatchableStoreTxnPut-4   	   30000	     58477 ns/op	    1537 B/op	      19 allocs/op
```
After:
```
BenchmarkWatchableStoreTxnPut-4        30000      45182 ns/op      1349 B/op        14 allocs/op
```
Profiling with `memprofilerate 1`.
Before:
```
ROUTINE ======================== github.com/coreos/etcd/mvcc.(*watchableStore).TxnEnd in /home/nekto0n/workspace/etcd/src/github.com/coreos/etcd/mvcc/watchable_store.go
   10.10MB    17.92MB (flat, cum) 32.66% of Total
         .          .    134: s.mu.Lock()
         .          .    135: return s.store.TxnBegin()
         .          .    136:}
         .          .    137:
         .          .    138:func (s *watchableStore) TxnEnd(txnID int64) error {
         .     1.39MB    139: err := s.store.TxnEnd(txnID)
         .          .    140: if err != nil {
         .          .    141:    return err
         .          .    142: }
         .          .    143:
    9.19MB     9.19MB    144: changes := s.getChanges()
         .          .    145: if len(changes) == 0 {
         .          .    146:    s.mu.Unlock()
         .          .    147:    return nil
         .          .    148: }
         .          .    149:
         .          .    150: rev := s.store.Rev()
  940.66kB   940.66kB    151: evs := make([]mvccpb.Event, len(changes))
         .          .    152: for i, change := range changes {
         .          .    153:    switch change.CreateRevision {
         .          .    154:    case 0:
         .          .    155:       evs[i] = mvccpb.Event{
         .          .    156:          Type: mvccpb.DELETE,
         .          .    157:          Kv:   &changes[i]}
         .          .    158:       evs[i].Kv.ModRevision = rev
         .          .    159:    default:
         .          .    160:       evs[i] = mvccpb.Event{
         .          .    161:          Type: mvccpb.PUT,
         .          .    162:          Kv:   &changes[i]}
         .          .    163:    }
         .          .    164: }
         .          .    165:
         .     6.43MB    166: s.notify(rev, evs)
         .          .    167: s.mu.Unlock()
         .          .    168:
         .          .    169: return nil
         .          .    170:}
         .          .    171:
```
After (allocations in `.notify()` are gone):
```
ROUTINE ======================== github.com/coreos/etcd/mvcc.(*watchableStore).TxnEnd in /home/nekto0n/workspace/etcd/src/github.com/coreos/etcd/mvcc/watchable_store.go
   13.46MB    15.31MB (flat, cum) 23.18% of Total
         .          .    134:	s.mu.Lock()
         .          .    135:	return s.store.TxnBegin()
         .          .    136:}
         .          .    137:
         .          .    138:func (s *watchableStore) TxnEnd(txnID int64) error {
         .     1.85MB    139:	err := s.store.TxnEnd(txnID)
         .          .    140:	if err != nil {
         .          .    141:		return err
         .          .    142:	}
         .          .    143:
   12.24MB    12.24MB    144:	changes := s.getChanges()
         .          .    145:	if len(changes) == 0 {
         .          .    146:		s.mu.Unlock()
         .          .    147:		return nil
         .          .    148:	}
         .          .    149:
         .          .    150:	rev := s.store.Rev()
    1.22MB     1.22MB    151:	evs := make([]mvccpb.Event, len(changes))
         .          .    152:	for i, change := range changes {
         .          .    153:		switch change.CreateRevision {
         .          .    154:		case 0:
         .          .    155:			evs[i] = mvccpb.Event{
         .          .    156:				Type: mvccpb.DELETE,
```

### CODE REVIEW

File: mvcc/watcher_group.go [hunk 1]
```diff
@@ -256,8 +260,11 @@ func (wg *watcherGroup) chooseAll(curRev, compactRev int64) int64 {
 // watcherSetByKey gets the set of watchers that receive events on the given key.
 func (wg *watcherGroup) watcherSetByKey(key string) watcherSet {
 	wkeys := wg.keyWatchers[key]
-	wranges := wg.ranges.Stab(adt.NewStringAffinePoint(key))
-
+	// Don't allocate intervals slice if we do not have any range watchers
+	var wranges []*adt.IntervalValue
+	if wg.ranges.Len() != 0 {
```
Comment (heyitsanthony) [inline mvcc/watcher_group.go]: Would this check be more useful in `IntervalTree.Stab`?

File: mvcc/watchable_store_bench_test.go [hunk 2]
```diff
@@ -39,6 +39,31 @@ func BenchmarkWatchableStorePut(b *testing.B) {
...
+	s := New(be, &lease.FakeLessor{}, &i)
+	defer cleanup(s, be, tmpPath)
+
+	// arbitrary number of bytes
+	bytesN := 64
+	keys := createBytesSlice(bytesN, b.N)
+	vals := createBytesSlice(bytesN, b.N)
+
+	b.ResetTimer()
+	b.ReportAllocs()
```
Comment (heyitsanthony) [inline mvcc/watchable_store_bench_test.go]: Add `ReportAllocs` to `BenchmarkWatchableStorePut` too? Probably useful to have the stats for both.

File: mvcc/watcher_group.go [hunk 3]
```diff
@@ -78,6 +78,10 @@ func (wb watcherBatch) add(w *watcher, ev mvccpb.Event) {
 // newWatcherBatch maps watchers to their matched events. It enables quick
 // events look up by watcher.
 func newWatcherBatch(wg *watcherGroup, evs []mvccpb.Event) watcherBatch {
+	// Don't allocate anything if we don't have any watchers.
+	if wg.size() == 0 {
```
Comment (heyitsanthony) [inline mvcc/watcher_group.go]: This same change is in #5316. I think that should get merged since it's in this one too. /cc @xiang90

Comment (xiang90) [inline mvcc/watcher_group.go]: merged. this needs a rebase.

Comment (nekto0n) [inline mvcc/watcher_group.go]: Oh, right, missed that PR. Actually, this particular change is enough to shave off allocations in `TxnEnd` without watches. Maybe that PR was not very convincing without benchmarks.

File: mvcc/watcher_group.go (hunk 1 above)
Comment (nekto0n) [inline mvcc/watcher_group.go]: I guess I'd better make another PR for this particular change?

**Comment by nekto0n** [comment]
Closing this one, will split it into separate PRs.
